File: backend/idhazh/contracts/feed_health.py
"""What every feed did, on every run.

One row per feed per run, appended to `state/feed-health/<YYYY-MM>.csv`. It is
written whether the run publishes or not, because the days a source is worth
measuring on are the days the run went badly.

Two readers depend on it. The plan stage reads the recent tail to decide
whether a feed has failed often enough to be skipped - a quarantine that lives
in the record, never in `config/sources.json`, so no run ever edits a file a
person owns. The published dashboard reads the same rows to show which sources
are healthy, which is the only way a reader can tell a quiet desk from a broken
one.

`FetchOutcome` lives here rather than in `idhazh.fetch` because it is now a
persisted vocabulary: it is a column in this ledger and a label on a published
page. `RobotsOutcome` is here for the same reason. Contracts are the bottom of
the dependency graph (CLAUDE.md section 4), so the shape a run writes cannot
live in the module that produces it.
"""

from __future__ import annotations

import hashlib
from enum import StrEnum
from typing import Any, ClassVar, Self

from pydantic import Field

from idhazh.contracts.base import (
    ChangelogEntry,
    Contract,
    DateStamp,
    RunId,
    Sha256,
    Slug,
    Timestamp,
)


def derive_endpoint_key(feed_url: str) -> str:
    """Which address a run asked, as one cell.

    The sha256 of the validated configured feed URL. It says what was asked
    without putting the address in a file the console reads, and it is what a
    later retirement is filed against - so a feed whose configured URL changes
    is a new endpoint with no inherited record.

    Unlike every other derived value here it cannot be rebuilt on read, because
    the row deliberately does not carry the URL. The writer derives it with this
    function and nothing else spells the arithmetic.
    """
    return hashlib.sha256(feed_url.encode("utf-8")).hexdigest()


class FetchOutcome(StrEnum):
    """Why one read of one address ended the way it did.

    Deliberately coarser than an HTTP status. What a later decision needs is
    whether the address is worth asking again, and 403 and 404 answer that the
    same way while 503 answers it differently.
    """

    OK = "ok"
    ROBOTS_DENIED = "robots_denied"
    BLOCKED = "blocked"
    PERMANENT = "permanent"
    TRANSIENT = "transient"
    # We did not ask - the feed was resting out a quarantine. A record, not a
    # measurement. It is written so a later run can tell "resting" apart from
    # "nobody has tried this feed yet", which is what lets a quarantine lift
    # itself instead of becoming a deletion nobody voted for.
    SKIPPED = "skipped"


class RobotsOutcome(StrEnum):
    """What `robots.txt` said about this address, as a decision and never a body.

    Three answers because three are actionable. A host that publishes no rules
    and a host whose rules allow us are the same permission, so they are one
    member; a refusal and a robots file we could not read are different, because
    the first is a publisher's stated policy and the second is our own failure
    to establish one. Neither of the last two says anything about whether the
    feed itself works, so neither is evidence about availability.
    """

    ALLOWED = "allowed"
    DENIED = "denied"
    # We could not establish permission - a 429, a 5xx, or a transport failure.
    # Unknown fails closed, so the target is not asked.
    UNREACHABLE = "unreachable"


#: Outcomes that count against a feed when quarantine is decided. A robots
#: refusal is not one of them: the source is working exactly as it asked to be
#: treated, and quarantining it would be us punishing a site for saying no.
FAILING_OUTCOMES: frozenset[FetchOutcome] = frozenset(
    {FetchOutcome.BLOCKED, FetchOutcome.PERMANENT, FetchOutcome.TRANSIENT}
)


class FeedHealthRow(Contract):
    """One feed, one run, one row."""

    __schema_stem__: ClassVar[str] = "feed-health-row"
    __changelog__: ClassVar[tuple[ChangelogEntry, ...]] = (
        ChangelogEntry(
            version="2026-09-02T20:00",
            change=(
                "endpoint_key, robots_outcome, robots_checked_at, robots_status and "
                "target_attempted appended, every one of them nullable. The committed "
                "shards are rewritten onto the wider header in the same commit by "
                "backend/utilities/migrate_feed_health.py, and every rewritten row "
                "carries five empty cells."
            ),
            why=(
                "The record could say what a feed answered and not which address was "
                "asked, whether the site allowed the request, or whether the request "
                "was made at all. A run cannot retire a dead address without knowing "
                "the address, and a robots refusal reads as a failed read without the "
                "permission beside it. Empty rather than backfilled: the configured URL "
                "may have moved since a row was written, and a guessed identity would "
                "retire the wrong address later."
            ),
        ),
        ChangelogEntry(
            version="2026-08-23T12:00",
            change="outcome gains `skipped`, for a run that held a quarantined feed back.",
            why=(
                "A quarantine that writes nothing can never lift: the failures that "
                "caused it stay the newest thing on record forever. A skip has to be "
                "recorded for the rest to end on its own."
            ),
        ),
        ChangelogEntry(
            version="2026-08-23",
            change="Initial shape: the feed, the run, the outcome and what it yielded.",
            why=(
                "A feed can go quiet for a week before anybody notices, and nothing "
                "recorded whether a source answered. A quarantine rule and a health "
                "dashboard both need a measurement that already exists."
            ),
        ),
    )

    run_id: RunId
    date: DateStamp
    feed_id: Slug
    checked_at: Timestamp
    outcome: FetchOutcome
    status: int | None = Field(
        default=None,
        ge=100,
        le=599,
        description="The HTTP status, when the request got far enough to have one.",
    )
    items: int = Field(
        default=0,
        ge=0,
        description=(
            "Candidates the feed yielded. Zero with an ok outcome is its own kind of "
            "broken - a feed that parses to nothing is not a feed that is working."
        ),
    )
    detail: str | None = Field(
        default=None,
        max_length=200,
        description="Our own one-line reason. Never the response body (Rule #11).",
    )
    # Everything below here was appended on 2026-09-02 and is empty on every row
    # written before that. A column is appended and never filed by meaning: a
    # cell inserted in the middle shifts every historical value one place right
    # under a reader that maps by position.
    endpoint_key: Sha256 | None = Field(
        default=None,
        description=(
            "The sha256 of the configured feed URL this run asked, from "
            "derive_endpoint_key. It identifies the address rather than the feed, so a "
            "feed whose URL is edited starts a fresh record. Empty on a row written "
            "before the column existed."
        ),
    )
    robots_outcome: RobotsOutcome | None = Field(
        default=None,
        description=(
            "What robots.txt said about this address. Empty on a row written before "
            "the column existed, which is not the same fact as allowed."
        ),
    )
    robots_checked_at: Timestamp | None = Field(
        default=None,
        description="When permission was established, so a recheck cadence has a clock.",
    )
    robots_status: int | None = Field(
        default=None,
        ge=100,
        le=599,
        description="The status robots.txt itself answered with, when there was one.",
    )
    target_attempted: bool | None = Field(
        default=None,
        description=(
            "Did this run request the feed address itself? False for a run that stopped "
            "at robots.txt or rested the feed, so evidence about the feed can be told "
            "apart from evidence about permission. The cell is spelled True or False, "
            "as every other boolean cell under state/ is."
        ),
    )

    @property
    def attempted(self) -> bool:
        """Did we actually ask this run? A quarantined feed was not asked."""
        return self.outcome is not FetchOutcome.SKIPPED

    @property
    def failing(self) -> bool:
        """Did this read count against the feed?

        A successful read that parsed to no entries counts. The most common way
        a feed dies is not a 500 - it is a silent reshape that still returns 200.
        A robots refusal never counts, and neither does a run we skipped: the
        first is a site working as it asked to be treated, and the second is a
        question we did not ask.
        """
        if self.outcome is FetchOutcome.OK:
            return self.items == 0
        return self.outcome in FAILING_OUTCOMES

    @classmethod
    def csv_columns(cls) -> tuple[str, ...]:
        """One definition, so a writer and a reader cannot disagree about the shape."""
        return tuple(cls.model_fields)

    def csv_row(self) -> dict[str, str]:
        """Every cell a string. An absent optional is an empty cell, not the word None."""
        payload = self.model_dump(mode="json")
        return {name: "" if payload[name] is None else str(payload[name]) for name in payload}

    @classmethod
    def from_csv_row(cls, row: dict[str, str]) -> Self:
        """The inverse. An empty cell is an absent value, never the empty string.

        A missing cell is absent too, so a shard written before a column existed
        reads with that column empty rather than raising. The guard that forces
        the migration is `ledger.require_matching_header`, and it sits on the
        write path where an unmigrated file would put cells under the wrong
        names.
        """
        payload: dict[str, Any] = {name: row.get(name, "") for name in cls.model_fields}
        for name in cls._absent_when_blank():
            if payload[name] == "":
                payload[name] = None
        return cls.model_validate(payload)

    @classmethod
    def _absent_when_blank(cls) -> tuple[str, ...]:
        """Every optional cell, derived rather than listed a second time.

        A field declared `default=None` is one this row can be missing. Reading
        that off the model means a column added later cannot be forgotten here
        and come back from the ledger as the string `""`.
        """
        return tuple(name for name, field in cls.model_fields.items() if field.default is None)

File: frontend/src/lib/feed-health.ts
/** The quarantine rule, restated on the read side, and the day roll-up under it.
 *
 * The console has to print the number the pipeline rested a feed on. That is a
 * run of failures ending at the newest read - not every failure the ledger ever
 * held. The two are different the moment a feed comes back: a source that
 * failed twelve times in July and answered this morning is healthy, and a total
 * printed beside a quarantine marker says the pipeline dropped it for a reason
 * it never used.
 *
 * So the rule lives here rather than in the page, in one place a test can drive
 * with rows it made up, and it is the same loop `discover.streak` runs over the
 * same evidence `discover.settled` leaves.
 * `backend/idhazh/contracts/feed_health.py` and `backend/idhazh/discover.py`
 * are the source of truth; this is a reader of the same ledger and it has to
 * agree with them.
 */

import { plural, shortDate } from './format';

/** What this module needs off a feed-health row, and nothing more. */
export interface FeedRead {
	date: string;
	runId: string;
	outcome: string;
	items: number;
}

/** The same set as `FAILING_OUTCOMES` in the contract. A robots refusal is not
 * one of them: the source is working exactly as it asked to be treated, and
 * resting it would be us punishing a site for saying no. */
const FAILING_OUTCOMES = new Set(['blocked', 'permanent', 'transient']);

/** The same set as `PRESERVING_OUTCOMES` in the contract: a streak runs through
 * these untouched. `robots_denied` is what the ledger writes for a refusal and
 * for a robots.txt we could not read, and availability does not care which. */
const PRESERVING_OUTCOMES = new Set(['robots_denied', 'skipped']);

/** Did this read count against the feed?
 *
 * A successful read that parsed to no entries counts. The most common way a
 * feed dies is not a 500 - it is a silent reshape that still returns 200.
 */
export function failing(row: FeedRead): boolean {
	if (row.outcome === 'ok') return row.items === 0;
	return FAILING_OUTCOMES.has(row.outcome);
}

/** Did the address itself come back carrying entries? The one result that
 * clears a streak, and the one that wins a repeated run. */
export function answered(row: FeedRead): boolean {
	return row.outcome === 'ok' && row.items > 0;
}

/** Does a streak of failures run through this row untouched?
 *
 * True for a rest and for a robots result. Neither one asked the feed whether
 * it still works, so neither may add a strike and neither may clear one.
 * Clearing on a refusal is the sharper error: a dead address would launder its
 * record every time the site said no.
 */
export function preserves(row: FeedRead): boolean {
	return PRESERVING_OUTCOMES.has(row.outcome);
}

/** We did not ask. A rest is a record, not a measurement. */
export function skipped(row: FeedRead): boolean {
	return row.outcome === 'skipped';
}

/** Oldest run first, which is the order the rules below read in.
 *
 * A date alone does not order five runs of one day, and a stable sort then
 * leaves the newest read as whichever row the shard happened to carry last.
 */
export function chronological<T extends FeedRead>(rows: readonly T[]): T[] {
	return [...rows].sort((a, b) => a.date.localeCompare(b.date) || a.runId.localeCompare(b.runId));
}

/** Failures in a row, ending at the newest read. `rows` are oldest run first.
 *
 * Four kinds of evidence and three effects, which is the whole availability
 * rule:
 *
 * - blocked, permanent, transient, and a success carrying no entries: one
 *   strike each. An empty answer costs the digest exactly what a refusal does.
 * - a robots result and a rest: neither adds a strike nor clears one. Neither
 *   one asked the feed whether it still works.
 * - a read that carried entries: the streak is over. The endpoint answers now,
 *   and the ledger keeps every old failure for the reliability record.
 *
 * This is `discover.streak` and it has to stay that loop.
 */
export function streak(rows: readonly FeedRead[]): number {
	let strikes = 0;
	for (let index = rows.length - 1; index >= 0; index -= 1) {
		const row = rows[index];
		if (preserves(row)) continue;
		if (!failing(row)) break;
		strikes += 1;
	}
	return strikes;
}

/** Is the pipeline resting this feed right now? The same rule as `_rests`.
 *
 * The rest ends on its own, or a bad afternoon becomes a permanent removal: a
 * feed skipped as many times as it was struck is asked again regardless. Both
 * counters read the one knob, because there is only one question here - how
 * much evidence is enough.
 */
export function resting(rows: readonly FeedRead[], after: number): boolean {
	let skips = 0;
	for (let index = rows.length - 1; index >= 0; index -= 1) {
		if (!skipped(rows[index])) break;
		skips += 1;
	}
	if (skips >= after) return false;
	return streak(rows) >= after;
}

/** One read, and which feed made it. */
export interface FeedRecord extends FeedRead {
	feedId: string;
}

/** A read the settlement can place: which feed, which run, and when we looked. */
export interface FeedEvent extends FeedRecord {
	checkedAt: string;
}

/** Does `later` replace `kept` as this run's one result for this feed?
 *
 * `contracts.feed_health.supersedes`, restated. A read that carried entries
 * wins, whichever row is newer: the attempt that got articles is the attempt
 * that happened. Between two rows that agree on that, the later `checkedAt`
 * wins. A tie leaves the row already held.
 */
function supersedes(later: FeedEvent, kept: FeedEvent): boolean {
	if (answered(later) !== answered(kept)) return answered(later);
	return later.checkedAt > kept.checkedAt;
}

/** One result per feed per run, in the order the rows arrived.
 *
 * A feed is read once in a run, so two rows under one run and feed are two
 * accounts of one event, and counting both counts a run twice. They exist
 * because a second attempt at a run cannot see what the first attempt pushed
 * after its checkout, and the union merge on `state/**\/*.csv` keeps both
 * lines rather than conflicting.
 *
 * `discover.settled` is the same reduction over the same rows. The page and the
 * run have to agree about how many times a feed failed, or the console
 * contradicts the pipeline that produced it.
 */
export function settled<T extends FeedEvent>(rows: readonly T[]): T[] {
	const kept = new Map<string, T>();
	for (const row of rows) {
		const key = `${row.runId}\u0000${row.feedId}`;
		const held = kept.get(key);
		if (held === undefined || supersedes(row, held)) kept.set(key, row);
	}
	return [...kept.values()];
}

/** What a feed did on one day, as the worst thing that happened to it.
 *
 * Four states rather than three: a source that said no in `robots.txt` neither
 * failed nor delivered, and drawing it as either would say something the ledger
 * does not.
 */
export type FeedDayOutcome = 'answered' | 'failed' | 'refused' | 'resting';

export interface FeedDay {
	date: string;
	outcome: FeedDayOutcome;
	/** The day and its whole tally, as a sentence. Colour is one signal and
	 * never the only one, and this is the other one. */
	label: string;
}

/** One square a day, oldest first, and the sentence under each.
 *
 * A day is drawn by its worst outcome, because the day an operator has to look
 * at is the day something failed. The label carries the whole tally, so a day
 * that mostly worked is never reported as a day that did not.
 */
export function feedDays(rows: readonly FeedRead[]): FeedDay[] {
	const byDay = new Map<string, FeedRead[]>();
	for (const row of rows) byDay.set(row.date, [...(byDay.get(row.date) ?? []), row]);

	return [...byDay]
		.map(([date, group]) => {
			const tally = {
				answered: group.filter(answered).length,
				failed: group.filter(failing).length,
				refused: group.filter((row) => row.outcome === 'robots_denied').length,
				resting: group.filter(skipped).length
			};
			const outcome: FeedDayOutcome =
				tally.failed > 0
					? 'failed'
					: tally.refused > 0
						? 'refused'
						: tally.answered > 0
							? 'answered'
							: 'resting';
			const parts = [
				tally.answered > 0 ? `${tally.answered} answered` : '',
				tally.failed > 0 ? `${tally.failed} failed` : '',
				tally.refused > 0 ? `${tally.refused} politely refused` : '',
				tally.resting > 0 ? `${tally.resting} not asked` : ''
			].filter(Boolean);
			return {
				date,
				outcome,
				label: `${shortDate(date)}, ${plural(group.length, 'run', 'runs')}: ${parts.join(', ')}.`
			};
		})
		.sort((a, b) => a.date.localeCompare(b.date));
}

/** What to print in the last-result cell.
 *
 * The ledger's own word for a feed that answered 200 with no entries is `ok`,
 * because that is what the fetch did. Printed raw it sits on the same row as a
 * count of fourteen failures and contradicts it, which is how a dead feed reads
 * as a healthy one.
 */
export function resultLabel(row: FeedRead): string {
	if (row.outcome === 'ok' && row.items === 0) return 'answered with nothing';
	return row.outcome;
}

/** How many feeds have never failed, out of how many the pipeline has read,
 * over how many runs.
 *
 * The console lists only the feeds that broke, which is the right list and half
 * an answer: four broken feeds out of eight is a collapse and four out of two
 * hundred is a Tuesday, and the page drew both identically. This is the
 * denominator, and it is a count rather than a share - the number an operator
 * acts on is how many are named below, not a percentage.
 */
export interface Reliability {
	/** Feed ids with at least one read of the feed and no failing one,
	 * alphabetically. Alphabetical because there is no order: a feed is read
	 * once a run, so every clean feed has the same record. */
	clean: string[];
	/** Feeds the pipeline has actually read. A rest and a robots answer are both
	 * absent from this count, and for one reason: neither of them asked the feed
	 * whether it still works, so neither can make it clean or broken. */
	checked: number;
	/** Feeds with at least one failing read. `clean.length + failed` is
	 * `checked`, always. */
	failed: number;
	/** Feeds on record that the pipeline has never actually read - held back by
	 * a rest or by a robots answer on every run. Alphabetically, for the same
	 * reason `clean` is.
	 *
	 * They are named rather than counted into either side because a source
	 * honouring its own `robots.txt` has not failed and has not delivered
	 * either. Until 2026-09-03 a refusal counted as an ask, so a source that has
	 * never once given us an article was reported as one that had never failed -
	 * measured over the committed ledger that day, 5 feeds of 184. That is the
	 * one number this section existed to get right. */
	ineligible: string[];
	/** Runs the ledger holds. This is the span "never failed" is read over, and
	 * a shallow record is why the page has a third sentence: two runs deep,
	 * "never failed" means "did not fail twice". */
	runs: number;
}

/** The whole record, not a window.
 *
 * The same span the streak beside each feed is read over, because the pipeline
 * rests on the whole count and not on a windowed one. Two spans in one section
 * is the defect the shared window exists to remove.
 */
export function reliability(rows: readonly FeedRecord[]): Reliability {
	const read = new Map<string, FeedRecord[]>();
	const seen = new Set<string>();
	for (const row of rows) {
		seen.add(row.feedId);
		// `preserves` is the same predicate the streak runs on, and it is the
		// right one here for the same reason: a rest and a robots answer are not
		// evidence about the address either way. Reusing it is what stops this
		// sentence and the quarantine disagreeing about what an ask is.
		if (preserves(row)) continue;
		read.set(row.feedId, [...(read.get(row.feedId) ?? []), row]);
	}

	const clean: string[] = [];
	let failed = 0;
	for (const [feedId, reads] of read) {
		if (reads.some(failing)) failed += 1;
		else clean.push(feedId);
	}

	const byName = (a: string, b: string) => a.localeCompare(b);
	return {
		clean: clean.sort(byName),
		checked: read.size,
		failed,
		ineligible: [...seen].filter((feedId) => !read.has(feedId)).sort(byName),
		runs: new Set(rows.map((row) => row.runId)).size
	};
}
